kernel/collection/array/array.c

```diff
@@ -0,0 +1,125 @@
+//*****************************************************************
+// This file is part of CosmOS                                    *
+// Copyright (C) 2020 Tom Everett                                 *
+// Released under the stated terms in the file LICENSE            *
+// See the file "LICENSE" in the source distribution for details  *
+// ****************************************************************
+
+#include <collection/array/array.h>
+
+#include <mm/mm.h>
+#include <panic/panic.h>
+#include <console/console.h>
+
+struct array* array_new(uint32_t size) {
+    struct array*  ret = (struct array*) kmalloc(sizeof(struct array));
+    ret->data = (void**) kmalloc(sizeof(void*)*size); 
+  //  kprintf("array data pointer: %llX\n", (uint64_t)ret->data);
+    ret->size = size;
+    for (uint32_t i=0; i<size;i++){
+        ret->data[i]=0;
+    }
+    return ret;
+}
+
+void array_delete(struct array* arr){
+    if (0!=arr){
+        if (0!=arr->data){
+            kfree(arr->data);
+        }
+        kfree(arr);
+    } else {
+        panic("null array passed to array_delete\n");
+    }
+}
+
+uint32_t array_size(struct array* arr){
+    if (0!=arr){
+        return arr->size;
+    } else {
+        panic("null array passed to array_size\n");
+    }   
+}
+
+void  array_set(struct array* arr, uint32_t position, void* value){
+     if (0!=arr){
+         if ((position>=0) &&(position< arr->size) ){
+             arr->data[position]=value;
+         } else {
+            kprintf("Index %llu",position);
+            panic("invalid array index passed to array_set\n");
+        }
+     } else {
+        panic("null array passed to array_set\n");
+    }  
+}
+
+void* array_get(struct array* arr,  uint32_t position){
+    if (0!=arr){
+        if ((position>=0) &&(position< arr->size) ){
+            return arr->data[position];
+        } else {
+            kprintf("Index %llu",position);
+            panic("invalid array index passed to array_get\n");
+        }  
+    } else {
+        panic("null array passed to array_get\n");
+    }  
+}
+
+void array_resize(struct array* arr, uint32_t size) {
+    if (0!=arr){
+        if (size >= arr->size){
+            /*
+            * TODO.  Should use krealloc here, but I think there is a bug in krealloc
```

Comment:
What are you seeing in krealloc?  I'll take a stab at it, I'm working on MMU stuff anyway.

Reply:
if I used krealloc, I get random crashes.  

So, I re-wrote the code to just allocate a new block of memory, copy the data and release the old block: it crashes too.

If you uncomment the line with kfree() on it, you will see that releasing that block of memory causes a crash.